pkg/ebpf/c/tracer-ebpf.c

```diff
@@ -501,8 +515,44 @@ static void cleanup_tcp_conn(struct pt_regs* ctx, conn_tuple_t* tup) {
         conn.conn_stats = *cst;
     }
 
-    // Send the connection data to the perf buffer
-    bpf_perf_event_output(ctx, &tcp_close_event, cpu, &conn, sizeof(conn));
+    // Batch TCP closed connections before generating a perf event
+    batch_t *batch_ptr = bpf_map_lookup_elem(&tcp_close_batch, &cpu);
+    if (batch_ptr == NULL) {
+        return;
+    }
+
+    // TODO: Can we turn this into a macro based on TCP_CLOSED_BATCH_SIZE?
+    switch (batch_ptr->pos) {
+    case 0:
+        batch_ptr->c0 = conn;
+        batch_ptr->pos++;
+        return;
+    case 1:
+        batch_ptr->c1 = conn;
+        batch_ptr->pos++;
+        return;
+    case 2:
+        batch_ptr->c2 = conn;
+        batch_ptr->pos++;
+        return;
+    case 3:
+        batch_ptr->c3 = conn;
+        batch_ptr->pos++;
+        return;
+    case 4:
+        batch_ptr->c4 = conn;
+        batch_ptr->pos++;
+        return;
+    case 5:
+        // We have a full batch, so we flush it to the perf buffer
+        batch_ptr->c5 = conn;
+        batch_ptr->pos = 0;
```

Comment:
Is this being zeroed out so that it be reused?

If so, is this racy?  If the batch pointer is being sent as the event, is it possible for new close events to start being added to the batch before userspace processes it?

Reply:
I assumed `bpf_perf_event_output` would copy the data into a buffer at the moment you flush it, but digging into the Kernel code, I only saw a data copy to a user space buffer when the perf ring is read. So indeed there could be a race.
Thinking about it though, you're allowed to write things that are allocated in the eBPF stack which don't outlive the `bpf_perf_event_output` call, so there has to be some magic I'm missing there 🤔 
In any case I can add some telemetry to validate that. Basically when we read a batch in userspace we can assume that if `pos != TCP_CLOSED_BATCH_SIZE-1` it means a data race hapenned and we lost a tcp_close event.

Reply:
I made the test I mentioned above and didn't detect any race on a HAProxy node processing 15k connections per second.

Reply:
👍

Were they being closed at that rate?

Reply:
Yup